### ISSUE TITLE
Unable To Get LocalUser on Docker Container Running Random UID

### ISSUE THREAD

**Issue by mplachter** [ronf/asyncssh#346]
getpaas.getuser() is used in the `connection.py` `prepare` method which causes an error to be thrown on containers running as a random UID.

```
    def prepare(self, last_config=None, config=(), reload=False,
                client_factory=None, client_version=(), host='', port=(),
                tunnel=(), family=(), local_addr=(), tcp_keepalive=(),
                kex_algs=(), encryption_algs=(), mac_algs=(),
                compression_algs=(), signature_algs=(), host_based_auth=(),
                public_key_auth=(), kbdint_auth=(), password_auth=(),
                x509_trusted_certs=(), x509_trusted_cert_paths=(),
                x509_purposes='secureShellServer', rekey_bytes=(),
                rekey_seconds=(), login_timeout=(), keepalive_interval=(),
                keepalive_count_max=(), known_hosts=(),
                server_host_key_algs=(), username=(), password=None,
                client_host_keysign=(), client_host_keys=None,
                client_host_certs=(), client_host=None, client_username=(),
                client_keys=(), client_certs=(), passphrase=None, gss_host=(),
                gss_kex=(), gss_auth=(), gss_delegate_creds=(),
                preferred_auth=(), agent_path=(), agent_identities=(),
                agent_forwarding=(), pkcs11_provider=(), pkcs11_pin=None):
        """Prepare client connection configuration options"""

        local_username = getpass.getuser()
  ```
  
We need to handle the KeyError from getpass.getuser() which under the hood calls  pwd.getpwuid() and handle the UID not found error from getpwuid.

`pwd.getpwuid(os.getuid())[0]KeyError: 'getpwuid(): uid not found: XXXXX`

**Comment by ronf** [comment]
Thanks for the report.

This value is used as the default username on the remote system if you don't specify one, as well as the username which appears as the client username when doing host-based auth. It is also made available as a match criteria when reading configuration from the OpenSSH-compatible "config" file.

From what I understand, looking up the username based on the UID with getpwuid() is a last resort, only after finding that none of LOGNAME, USER, LNAME, and USERNAME are set. So, setting any of these values in the environment might be one way to work around this problem.

What would you like to see happen in this case? It wouldn't be hard to set this to something like the empty string if none of the environment variables are set and the reverse lookup of the UID doesn't work. However, if the caller doesn't specify the username explicitly in that case, I'm not sure what username could be passed in the connect request, or used in the other places I mentioned above.

**Comment by mplachter** [comment]
So the way i see it is we have a couple of different paths.

1. Leave it the way it is but handle the error gracefully and display to the USER what needs to happen, IE please set a user name for one of these environment variables.

```
class NoSetUser(Exception):
    pass
....

try:
    local_username = getpass.getuser()
except KeyError:
   raise NoSetUser("No Local User Fund please set one of the env vars 'LOGNAME', 'USER', 'LNAME', or 'USERNAME'")
```


2. Handle a random UID user + no environment variables set for LOGNAME, USER, LNAME, USERNAME. I believe the best thing for this would be just to set local_username to a blank or even valued string like

```
local_username = `u_asyncssh_client` 


try:
    local_username = getpass.getuser()
except KeyError:
   pass
 
```

I'm fine with either path.

**Comment by ronf** [comment]
Thanks for your thoughts on this.

I did a little digging into the OpenSSH code, and it appears that it currently exits very early on with an error in this case:

```c
        /* Get user data. */
        pw = getpwuid(getuid());
        if (!pw) {
                logit("No user exists for uid %lu", (u_long)getuid());
                exit(255);
        }
```

That said, the fact that you ran into this suggests to me that there may be legitimate cases where AsyncSSH is started up as a UID which has no corresponding password entry and with an environment that doesn't set a username, and if the client code explicitly sets a remote username and doesn't rely on matching on the local username when reading config, there's no reason not to allow the connection to proceed.

I'm currently thinking of going with one of the following two options when getpass.getuser() fails:

a. Set the local username to 'unknown'. If the remote username is not explicitly set, allow this 'unknown' value for the local username to also be used as the remote username, or in the case of host-based auth as the client username sent in the userauth request. This is basically your option 2.

b. Set the local username to the empty string, but raise an ValueError with a message like 'Username not set' if the remote username (or client username in host-based auth) is not explicitly set by the caller. This is a variation of your option 1 but only leading to a failure when the username wasn't provided as an explicit argument.

Option (a) is definitely simpler, but it would mean that servers might see this 'unknown' username in the auth request, which in most cases would probably turn into an auth failure. This might be confusing to the client if the caller meant to set a username and forgot or thought the container was setting one but it turned out it didn't.

**Comment by mplachter** [comment]
Thank you for your quick turn around on this.

Speaking as a consumer of this project I personally like option `b`.

Also if we raise the error maybe raising an error that tells the consumer what they need to do to fix it. IE pass a username or set one of the env variables would be extremely helpful.

If you allow MRs I could take this on.

Let me know \0.

Thanks.

**Comment by ronf** [comment]
Thanks for the offer. I had some time to look at this again today, though, and have provided a fix in commit c59b6cc in the "develop" branch.

In the end, I decided that spreading the check out to where the remote username or client username is set wasn't worth the extra complexity it introduced. Instead, I kept the change local to where getpass.getuser() was called (more like your option 1), adding a more informative error message.

I raised ValueError here rather than creating a custom exception, as this doesn't seem like something that would need to be caught and handled separately. A caller should know before making the connect() call whether the environment will be providing a local username or not, and the main goal here is just to let them know what went wrong if it doesn't, not to catch it at runtime.

Thanks for the report -- this will be included in the next release.
